src/journal/cat.c

```diff
@@ -123,14 +142,23 @@ static int run(int argc, char *argv[]) {
         if (r <= 0)
                 return r;
 
-        fd = sd_journal_stream_fd(arg_identifier, arg_priority, arg_level_prefix);
-        if (fd < 0)
-                return log_error_errno(fd, "Failed to create stream fd: %m");
+        outfd = sd_journal_stream_fd(arg_identifier, arg_priority, arg_level_prefix);
+        if (outfd < 0)
+                return log_error_errno(outfd, "Failed to create stream fd: %m");
+
+        if (arg_stderr_priority == arg_priority) {
+                errfd = outfd;
```

Comment:
I don't think it's dangerous.  When this condition runs, the behavior of the code is identical to what it was before the change.  Specifically the call to `rearrange_stdio` was `rearrange_stdio(STDIN_FILENO, fd, fd)` - it can handle taking the same descriptor for both stdout and stderr.  It has logic to handle closing the fds properly even when some are the same fd.

The other place that errfd is used is `TAKE_FD`.  In the usage there I believe it is equivalent to `errfd = -1;`.  I think it might make more sense to actually use the result of `TAKE_FD`, further justifying the use of that macro.  So, now the call is `rearrange_stdio(STDIN_FILENO, TAKE_FD(outfd), TAKE_FD(errfd));` - does that seem good?

Reply:
so, i think @yuwata has a point in general though in this specific case this wouldn't matter too much. we use `_cleanup_close_` for the fd vars, so that the fd is automatically closed when the function is exited. Thus, if you just assign the fd to the var here, in theory it could be closed twice. Now, in reality that's not gonna happen, as there is only one path out right now further down, and in that case the variables are invalidated already, so nothing bad can happen. However code lives, and sooner or later more code might get added in the middle, with additional error paths, and as soon as that happens things will break, as then suddenly the same fd gets closed twice. Let's prepare a bit for potential later code additions and do the clean-up paths correctly right away.

My proposal: simply leave errfd initialized to -1 in case the same fds shall be used, and then use the following to call rearrange_stdio:

```c
r = rearrange_stdio(STDIN_FILENO, outfd, err_fd < 0 ? outfd : errfd); /* Invalidates fd on succcess + error! */
TAKE_FD(outfd);
TAKE_FD(errfd);
```

or something like that...